packages/base/src/Internal/Static.hs

```diff
@@ -35,13 +36,34 @@ import Data.Proxy(Proxy)
 import Foreign.Storable(Storable)
 import Text.Printf
 
+import Data.Binary
+import GHC.Generics (Generic)
+import Data.Proxy (Proxy(..))
+
 --------------------------------------------------------------------------------
 
 type ℝ = Double
 type ℂ = Complex Double
 
 newtype Dim (n :: Nat) t = Dim t
-  deriving Show
+  deriving (Show, Generic)
+
+instance Binary a => Binary (Complex a)
+  where
+    put (r :+ i) = put (r, i)
+    get = (\(r,i) -> r :+ i) <$> get
+
+instance (KnownNat n, Binary a) => Binary (Dim n a) where
+  get = do
+    k <- get
+    let n = natVal (Proxy :: Proxy n)
+    if n == k
+      then Dim <$> get
+      else fail ("Expected dimension " ++ (show n) ++ ", but found dimension " ++ (show k))
```

Comment:
This dimension check still isn't completely safe---it check that the dimension that the programmer expects is equal to the dimension that the binary data claims to have---but the binary data could be lying. Maybe we should actually check the size of the parsed matrix/vector.